CA-388295: Fix bugs from py2->py3 conversion in perfmon and hfx_filename

stephenchengCloud opened the pull request:
- Fix py2->py3 bugs in perfmon and hfx_filename
- Added unit test for hfx_filename
- XenRT test passed: 192908 Ring3 BST + BVT 
- Scanned by pytype manually

stephenchengCloud commented on scripts/hfx_filename:
```diff
@@ -33,11 +33,11 @@ def rpc(session_id, request):
-          #print "%s\r\n" % h,
-        s.send(request)
+          s.send((h + "\r\n").encode('utf-8'))
+          #print ((h + "\r\n").encode('utf-8'))
```
I noticed that.
This script has a lot of prints, even the main funtion's output is print.
I assume this script is used to output to the terminal, so keep these print as they are.

duobei replied:
So we could delete all those lines of `#print` since those are used for debugging in the dev.

stephenchengCloud replied:
Delete unused comments.
Code updated.

stephenchengCloud commented on scripts/test_hfx_filename.py:
```diff
@@ -0,0 +1,105 @@
+        body = hfx_filename.rpc(session_id, request)
+        
+        # Assert that the socket methods were called as expected
+        expected_data_1 = b"POST /remote_db_access?session_id=0 HTTP/1.0\r\n"
+        expected_data_2 = b"Connection:close\r\n"
+        expected_data_3 = b"content-length:14\r\n"
+        expected_data_4 = b"\r\n"
+        expected_data_5 = b"socket request"
+        mock_connected_socket.send.assert_has_calls([
+            call(expected_data_1),
+            call(expected_data_2),
+            call(expected_data_3),
+            call(expected_data_4),
+            call(expected_data_5)
+        ])
```
Thx, code updated.

liulinC commented on scripts/test_hfx_filename.py:
```diff
@@ -0,0 +1,101 @@
+# mock modules to avoid dependencies
+sys.modules["XenAPI"] = MagicMock()
+
+def import_from_file(module_name, file_path):
```
This function should be common

stephenchengCloud replied:
Code updated!

liulinC commented on scripts/test_hfx_filename.py:
```diff
@@ -0,0 +1,101 @@
+# unittest for hfx_filename
+
+import unittest
+from mock import MagicMock, patch, call
```
is the unittest in the CI,  still using the 3rd party mock @acefei

liulinC commented:
@stephenchengCloud  Another issue may not related to your update.

Regarding `raise "Unable to parse string response"` (and some others), both py2 and py3 gives me
```TypeError: exceptions must be old-style classes or derived from BaseException, not str```

stephenchengCloud commented:
> @stephenchengCloud Another issue may not related to your update.
> 
> Regarding `raise "Unable to parse string response"` (and some others), both py2 and py3 gives me `TypeError: exceptions must be old-style classes or derived from BaseException, not str`

This is the error handling of this script (see line 56): if the input is not started with "<value><array><data><value>success</value><value>" for parse_string(), this error will be raised.
You probably run this script with the wrong input. I've already tested this function in the unittest.

stephenchengCloud commented:
XenRT test passed: 193536 Ring3 BST + BVT

freddy77 commented on scripts/hfx_filename:
```diff
@@ -31,15 +31,11 @@ def rpc(session_id, request):
           "content-length:%d" % (len(request)),
```
Request should be converted to binary before getting the length.

stephenchengCloud replied:
Thank you for your review.
Actually the unittest tested the function, but didn't consider the non-ascii inupt.
I added a test case for international characters so that they can be covered by the unittest:

> AssertionError: Calls not found.
> Expected: [call(b'POST /remote_db_access?session_id=0 HTTP/1.0\r\n'),
>  call(b'Connection:close\r\n'),
>  call(b'content-length:9\r\n'),
>  call(b'\r\n'),
>  call(b'socket \xe8\xaf\xb7\xe6\xb1\x82')]
> Actual: [call(b'POST /remote_db_access?session_id=0 HTTP/1.0\r\n'),
>  call(b'Connection:close\r\n'),
>  call(b'content-length:13\r\n'),
>  call(b'\r\n'),
>  call(b'socket \xe8\xaf\xb7\xe6\xb1\x82')]

freddy77 commented on scripts/import_file.py:
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env python3
```
This conflict with the fact that you check for python 2 in `import_from_file`.

liulinC replied:
This file is never run directly, but imported by other modules as a lib.
That is the reason why it support both py2 and py3 for compatiblity.
But it is still good to have this line removed.

stephenchengCloud replied:
This py2 check was added for the CI python2 pytest.
Will remove the shebang

freddy77 commented on scripts/import_file.py:
```diff
@@ -0,0 +1,27 @@
+    # Only for python3
+    if sys.version_info.major == 2:
+        return None
+    else:
```
You can avoid the `else` and reduce indentation

freddy77 commented on scripts/test_hfx_filename.py:
```diff
@@ -0,0 +1,107 @@
+        return_txt = hfx_filename.parse_string(txt)
+        self.assertEqual(expected_txt, return_txt)
+        
+        
```
Why not adding
```python
if __name__ == '__main__':
    unittest.main()
```
lines and make this script executable? It's working, not requiring extra dependencies or settings. It also will make shellbang line meaningful.

stephenchengCloud replied:
This module is used for importing non-'.py' files in the test, it never runs directly.
I'll remove the shebang.

psafont commented:
There's a warning in the test output regarding perfmon, can this be fixed as well, please?

```
=============================== warnings summary ===============================
scripts/perfmon:383
  /__w/xen-api/xen-api/scripts/perfmon:383: DeprecationWarning: invalid escape sequence '\d'
    memdict = dict([(k.strip(), float(re.search('\d+', v.strip()).group(0))) for (k,v) in memdict])
```

bernhardkaindl commented on scripts/perfmon:
```diff
@@ -1067,7 +1067,7 @@ def main():
 
             # Remove any monitors for VMs no longer listed in rrd_updates page
-            for uuid in vm_mon_lookup:
+            for uuid in list(vm_mon_lookup.keys()):
```
Please add a comment for readers to make it clear why list(dict_var.keys()) is needed here:
```suggestion
            # we use .pop() inside the loop, use list(dict_var.keys()):
            for uuid in list(vm_mon_lookup.keys()):
```

bernhardkaindl commented on scripts/perfmon:
```diff
@@ -1103,7 +1103,7 @@ def main():
 
             # Remove monitors for SRs no longer listed in the rrd_updates page
-            for uuid in sr_mon_lookup:
+            for uuid in list(sr_mon_lookup.keys()):
```
Please add a comment for readers to make it clear why list(dict_var.keys()) is needed here:
```suggestion
            # we use .pop() inside the loop, use list(dict_var.keys()):
            for uuid in list(sr_mon_lookup.keys()):
```

bernhardkaindl reviewed (changes requested):
Please add comments that clarify for readers (and future changes) that `list(dict_varible.keys())` is needed because the dict_variable is changed in the loops.

It helps others to understand the change during review and other developers in the future.

The comment is very much needed:
In the previous PR and all of us did not see it during review, and you yourself did not see it.

I added two suggestions that add a very short comment to each of them!

I updated the PR title to include CA-388295 with clear information that this fixes Py3 bugs.

bernhardkaindl reviewed (changes requested):
Thanks for adding the test for the hfx_filename.rpc() function.

I applied adding the comments that I requested to be added as well.

While the scripts are too large to be verified during a simple review, pytype errors are now fixed with it.

Like Pau mentioned above, please also fix te unsupported escape sequence warning printed when executing the code and by pyright (it is the backend of the official Python extension of Visual Studio Code), which you can install using `pip install pyright`:

```py
pyright scripts/perfmon
scripts/perfmon:383:54 - warning: Unsupported escape sequence in string literal (reportInvalidStringEscapeSequence)
(among many other messages)
```

stephenchengCloud commented:
- Fixed `Unsupported escape sequence` issue.
- Added comment on bug fix
- Squash the commits

psafont commented:
Waiting for bernhard to rereview the changes.

Is this passing tests that previously failed? Can we inspect the results?

stephenchengCloud commented:
> Waiting for bernhard to rereview the changes.
> 
> Is this passing tests that previously failed? Can we inspect the results?

@bernhardkaindl Could you review this? Thanks.
@psafont Regarding ''Is this passing tests that previously failed? Can we inspect the results?", I would say, these bugs were not covered by the previous test. There're three kinds of test and they all have limitation.
- pytype. It can not find bugs like `str-byte issue in sockect.recv()' and `changed size during iteration`.
- XenRT test. We alwasy ran XenRT tests before raise a PR. The previous PR were merged 3 weeks ago, but so far, most bugs in this PR were not triggered by any XenRT suite.
- Unit test. This is the most effective way for testing. That's why I added unit test in this PR. But even after I added unit test, @freddy77 still found a `str-byte issue in len(data)`.  Normally we mock the input and output of socket/subprocess in the unit test and that's the very places where problems are most likely to occur during py2->py3 migration.
Anyway, we should test as much as possible. I've learnt my lessons during these py3 migaration work and thank everyone who pointed out my mistakes.

stephenchengCloud commented:
1. Will fix the warnings by pylint check later

2. The issue described in CA-388295 has been fixed. Tested many times.There are no `RuntimeError: dictionary changed size during iteration` in the user.log (Job 3918881/3919173/3919778/3920228/3921218)

3. During testing, found another issue, took some time to fix and test.
It's about the Exception handling differences between py2 and py3
Please check the fix in this commit [800f97c](https://github.com/xapi-project/xen-api/pull/5373/commits/800f97c15d115dfdd76340072fd1bb303bf68952)

 **Exception handling changes in python3:**
- IOError and socket.error now have been merged into OSError.
See 
https://docs.python.org/3/library/exceptions.html#IOError
https://docs.python.org/3/library/socket.html#exceptions

- HTTPError doesn't have content in `args`. So we can't use `e.args[0]`. That's where the issue occured.
> Jan 30 08:00:49 fibreuk-01-04d-NIC1 /perfmon: PERFMON:   File "/opt/xensource/bin/perfmon", line 1131, in main#012    if e.args[0] == 111:
> Jan 30 08:00:49 fibreuk-01-04d-NIC1 /perfmon: PERFMON: IndexError: tuple index out of range

- Difference exception classes:
The inheritance relationship among these classes is as follows:
urllib.error.HTTPError <- urllib.error.URLError <- [OSError](https://docs.python.org/3/library/exceptions.html#OSError)
see :https://docs.python.org/3/library/urllib.error.html
ConnectionRefusedError -> ConnectionError -> OSError
see https://docs.python.org/3/library/exceptions.html#ConnectionRefusedError

4. Test results
The exceptions were caught correctly.
Below are logs from both python 2 and python 3:

user.log (job: 3919173) with python2 version perfmon:
> 	Line  11: Jan 30 10:24:46 fibreuk-01-01d /perfmon: PERFMON: caught socket.error: (2 No such file or directory) - restarting XAPI session
> 		Line  60: Jan 30 11:28:16 fibreuk-01-01d /perfmon: PERFMON: Caught signal 15 - exiting
> 		Line  72: Jan 30 11:31:43 fibreuk-01-01d-NIC1 perfmon: PERFMON: caught socket.error: (111 Connection refused) - restarting XAPI session
> 	  C:\Users\stephenche\Downloads\25437975-py2-perfmon\fibreuk-01-04d\user.log (8 hits)
> 		Line  11: Jan 30 10:25:15 fibreuk-01-04d /perfmon: PERFMON: caught socket.error: (2 No such file or directory) - restarting XAPI session
> 		Line  65: Jan 30 11:15:31 fibreuk-01-04d /perfmon: PERFMON: Caught signal 15 - exiting
> 		Line 113: Jan 30 11:19:00 fibreuk-01-04d-NIC1 perfmon: PERFMON: caught socket.error: (111 Connection refused) - restarting XAPI session
> 		Line 355: Jan 30 11:31:59 fibreuk-01-04d-NIC1 perfmon: PERFMON: caught IOError: (http error 401 Unauthorised content-length: 96#015#012content-type: text/html#015#012connection: close#015#012cache-control: no-cache, no-store#015#012WWW-Authenticate: Basic realm="rrd_updates"#015#012) - restarting XAPI session
	
user.log(job: 3921218) with python 3 version perfmon:
> Line  15: Feb  1 00:59:00 cl14-01 /perfmon: PERFMON: caught socket.error: (2 No such file or directory) - restarting XAPI session
> 	Line  72: Feb  1 01:16:11 cl14-01 /perfmon: PERFMON: Caught signal 15 - exiting
> 	Line 126: Feb  1 01:20:11 cl14-01-NIC2 /perfmon: PERFMON: caught connection refused error: ([Errno 111] Connection refused) - restarting XAPI session
> 	Line 360: Feb  1 01:33:38 cl14-01-NIC2 /perfmon: PERFMON: caught http.error: (HTTP Error 401: Unauthorised) - restarting XAPI session
> 	Line 382: Feb  1 01:43:21 cl14-01-NIC2 /perfmon: PERFMON: Caught signal 15 - exiting
> 	Line 434: Feb  1 01:47:15 cl14-01-NIC2 /perfmon: PERFMON: caught connection refused error: ([Errno 111] Connection refused) - restarting XAPI session
> 	Line 678: Feb  1 02:00:32 cl14-01-NIC2 /perfmon: PERFMON: caught http.error: (HTTP Error 401: Unauthorised) - restarting XAPI session
> Search "perfmon" (10 hits in 1 file of 1 searched)
>   C:\Users\stephenche\Downloads\3921218-py3-perfmon\cl14-03\user.log (10 hits)
> 	Line  15: Feb  1 00:58:33 cl14-03 /perfmon: PERFMON: caught socket.error: (2 No such file or directory) - restarting XAPI session
> 	Line  66: Feb  1 01:29:20 cl14-03 /perfmon: PERFMON: Caught signal 15 - exiting
> 	Line 120: Feb  1 01:33:24 cl14-03-NIC2 /perfmon: PERFMON: caught connection refused error: ([Errno 111] Connection refused) - restarting XAPI session
> 	Line 377: Feb  1 01:56:02 cl14-03-NIC2 /perfmon: PERFMON: Caught signal 15 - exiting
> 	Line 393: Feb  1 02:00:14 cl14-03 /perfmon: PERFMON: caught connection refused error: ([Errno 111] Connection refused) - restarting XAPI session

acefei commented:
> 1. Fixed the warnings by pylint and pytype
>    Some of the perfmon pytype warnings are false warnings:
>    Could you please check this? Thanks. @bernhardkaindl
> 
> ```
>  Warning: No attribute 'alarm_auto_inhibit_period' on VariableState
>   Warning: No attribute 'alarm_trigger_period' on VariableState
>   pytype_reporter: Progress: 20/28: check echo
>   pytype_reporter: Progress: 21/28: check scripts.xc
>   pytype_reporter: Progress: 22/28: check scripts.print-custom-templates
>   pytype_reporter: Progress: 23/28: check renameif
>   pytype_reporter: Progress: 24/28: check lvhd-api-test
>   pytype_reporter: Progress: 25/28: check provision
>   Warning: No attribute 'details' on Exception
> pytype_reporter:   No Node.TEXT_NODE in module xml.dom.minidom, referenced from 'xml.dom.expatbuilder'
> pytype_reporter:   No Node.TEXT_NODE in module xml.dom.minidom, referenced from 'xml.dom.expatbuilder'
> pytype_reporter: Progress: 26/28: check scripts.xe-reset-networking
> pytype_reporter: Progress: 27/28: check scripts.test_hfx_filename
> pytype_reporter: Progress: 28/28: check scripts.yum-plugins.accesstoken
> pytype for all files which are expected to pass
> Error: Process completed with exit code 3.
> ```
> 
> 2. The issue described in CA-388295 has been fixed. Tested many times.There are no `RuntimeError: dictionary changed size during iteration` in the user.log (Job 3918881/3919173/3919778/3920228/3921218)
> 3. During testing, found another issue, took some time to fix and test.
>    It's about the Exception handling differences between py2 and py3
>    Please check the fix in this commit [a98ea40](https://github.com/xapi-project/xen-api/commit/a98ea407e1c0125813571f4bdef8c5d5dce5239b)
> 
> **Exception handling changes in python3:**
> 
> * IOError and socket.error now have been merged into OSError.
>   See
>   https://docs.python.org/3/library/exceptions.html#IOError
>   https://docs.python.org/3/library/socket.html#exceptions
> * HTTPError doesn't have content in `args`. So we can't use `e.args[0]`. That's where the issue occured.
> 
> > Jan 30 08:00:49 fibreuk-01-04d-NIC1 /perfmon: PERFMON:   File "/opt/xensource/bin/perfmon", line 1131, in main#012    if e.args[0] == 111:
> > Jan 30 08:00:49 fibreuk-01-04d-NIC1 /perfmon: PERFMON: IndexError: tuple index out of range
> 
> * Difference exception classes:
>   The inheritance relationship among these classes is as follows:
>   urllib.error.HTTPError <- urllib.error.URLError <- [OSError](https://docs.python.org/3/library/exceptions.html#OSError)
>   see :https://docs.python.org/3/library/urllib.error.html
>   ConnectionRefusedError <- ConnectionError <- OSError
>   see https://docs.python.org/3/library/exceptions.html#ConnectionRefusedError
> 
> 4. Test results
>    The exceptions were caught correctly.
>    Below are logs from both python 2 and python 3:
> 
> user.log (job: 3919173) with python2 version perfmon:
> 
> > Line  11: Jan 30 10:24:46 fibreuk-01-01d /perfmon: PERFMON: caught socket.error: (2 No such file or directory) - restarting XAPI session
> > Line  60: Jan 30 11:28:16 fibreuk-01-01d /perfmon: PERFMON: Caught signal 15 - exiting
> > Line  72: Jan 30 11:31:43 fibreuk-01-01d-NIC1 perfmon: PERFMON: caught socket.error: (111 Connection refused) - restarting XAPI session
> > C:\Users\stephenche\Downloads\25437975-py2-perfmon\fibreuk-01-04d\user.log (8 hits)
> > Line  11: Jan 30 10:25:15 fibreuk-01-04d /perfmon: PERFMON: caught socket.error: (2 No such file or directory) - restarting XAPI session
> > Line  65: Jan 30 11:15:31 fibreuk-01-04d /perfmon: PERFMON: Caught signal 15 - exiting
> > Line 113: Jan 30 11:19:00 fibreuk-01-04d-NIC1 perfmon: PERFMON: caught socket.error: (111 Connection refused) - restarting XAPI session
> > Line 355: Jan 30 11:31:59 fibreuk-01-04d-NIC1 perfmon: PERFMON: caught IOError: (http error 401 Unauthorised content-length: 96#015#012content-type: text/html#015#012connection: close#015#012cache-control: no-cache, no-store#015#012WWW-Authenticate: Basic realm="rrd_updates"#15#012) - restarting XAPI session
> 
> user.log(job: 3921218) with python 3 version perfmon:
> 
> > Line  15: Feb  1 00:59:00 cl14-01 /perfmon: PERFMON: caught socket.error: (2 No such file or directory) - restarting XAPI session
> > Line  72: Feb  1 01:16:11 cl14-01 /perfmon: PERFMON: Caught signal 15 - exiting
> > Line 126: Feb  1 01:20:11 cl14-01-NIC2 /perfmon: PERFMON: caught connection refused error: ([Errno 111] Connection refused) - restarting XAPI session
> > Line 360: Feb  1 01:33:38 cl14-01-NIC2 /perfmon: PERFMON: caught http.error: (HTTP Error 401: Unauthorised) - restarting XAPI session
> > Line 382: Feb  1 01:43:21 cl14-01-NIC2 /perfmon: PERFMON: Caught signal 15 - exiting
> > Line 434: Feb  1 01:47:15 cl14-01-NIC2 /perfmon: PERFMON: caught connection refused error: ([Errno 111] Connection refused) - restarting XAPI session
> > Line 678: Feb  1 02:00:32 cl14-01-NIC2 /perfmon: PERFMON: caught http.error: (HTTP Error 401: Unauthorised) - restarting XAPI session
> > Search "perfmon" (10 hits in 1 file of 1 searched)
> > C:\Users\stephenche\Downloads\3921218-py3-perfmon\cl14-03\user.log (10 hits)
> > Line  15: Feb  1 00:58:33 cl14-03 /perfmon: PERFMON: caught socket.error: (2 No such file or directory) - restarting XAPI session
> > Line  66: Feb  1 01:29:20 cl14-03 /perfmon: PERFMON: Caught signal 15 - exiting
> > Line 120: Feb  1 01:33:24 cl14-03-NIC2 /perfmon: PERFMON: caught connection refused error: ([Errno 111] Connection refused) - restarting XAPI session
> > Line 377: Feb  1 01:56:02 cl14-03-NIC2 /perfmon: PERFMON: Caught signal 15 - exiting
> > Line 393: Feb  1 02:00:14 cl14-03 /perfmon: PERFMON: caught connection refused error: ([Errno 111] Connection refused) - restarting XAPI session

I think @bernhardkaindl has already demonstrated how to handle this warning through his actions. see https://github.com/xapi-project/xen-api/pull/5418/files#diff-e0b8a152da73af5c15f34d1fd7200f74f77ef38c4197c77ac2abeb765e78f99aR249

edwintorok commented:
Please change the target of the PR to the py3 feature branch.

liulinC commented:
> Please change the target of the PR to the py3 feature branch.

This is agreed with @psafont , bugfix goes to master,  and the issues already there. 
new py2->py3 should goes to feature branch for enough test,  to ensure no py2->py3 issues raised.

edwintorok commented on scripts/hfx_filename:
```diff
@@ -51,6 +51,7 @@ def rpc(session_id, request):
         s.close()
 
 def parse_string(txt):
+    assert txt is not None
```
Does this fix the original problem? It looks like 'rpc' may return on a non-200 error code, without raising an exception, and it'd instead fail on this assert here then.

stephenchengCloud replied:
This is not for the original problem.
The original change is small. Much of the code if for passing the pytype and pylint check added recently.
This is for passing the pytype.

edwintorok commented on scripts/perfmon:
```diff
@@ -379,8 +381,15 @@ def get_percent_mem_usage(ignored):
+            memtype = item[0].strip()
+            size = item[1].strip()
+            match = re.search(r'\d+', size.strip())
+            assert match
```
If there is no match this will raise an assertion error, it might be better to raise an error that gives you more information on what went wrong: in this case including the original `size` that we tried to match with the regex might be useful, so we see what it contains instead of digits.

bernhardkaindl replied:
- This change is not covered by tests.
  When installing and configuring the Codecov plugin for Chrome- or Firefox-based browsers, a red bar is shown in front of these lines. I recommend submitters to install and configure it (more info later, I'm in a meeting now) or check the code coverage on the Link given by the Codecov.io
- We also need to run mypy checks
- Since there are so many unresolved comments, and many changes, I think it's better to revert the Python3 changes as a fix, and move this to the Python3 branch instead to give this more time for review.

stephenchengCloud replied:
This is also for passing the pytype check.

acefei replied:
Agree with @edwintorok.
I'm assuming your modification is inspired by @bernhardkaindl's draft PR (https://github.com/xapi-project/xen-api/pull/5365/files#r1458703889) to avoid pytype errors. However, it would be preferable to make more elegant adjustments to display insightful exceptions.

edwintorok commented on scripts/perfmon:
```diff
@@ -500,7 +503,10 @@ class Variable(VariableConfig, VariableState):
+            # reset when reactivating
+            self.value = None
+            self.timeof_last_alarm = time.time() - self.alarm_auto_inhibit_period
+            self.trigger_down_counter = self.alarm_trigger_period
```
this duplicates code that previously wasn't duplicated.

stephenchengCloud replied:
Also for passing the pytype check.
The original code is a class, and pytype complained that the class has no `alarm_auto_inhibit_period` and `alarm_trigger_period` attributes.

stephenchengCloud commented:
OK. I will revert the python3 changes for hfx_filename and perfmon. @bernhardkaindl @psafont @edwintorok @liulinC @acefei 

For this PR:
1. The original fix is small, just a few lines changed, and I tested the fix both by XenRT and manual test.
2. Most of the changes are Unit test / pylint / pytype fix.

liulinC commented:
> OK. I will revert the python3 changes for hfx_filename and perfmon. @bernhardkaindl @psafont @edwintorok @liulinC @acefei
> 
> For this PR:
> 
> 1. The original fix is small, just a few lines changed, and I tested the fix both by XenRT and manual test.
> 2. Most of the changes are Unit test / pylint / pytype fix.

@psafont @edwintorok  @bernhardkaindl @acefei @stephenchengCloud 

I am very very very disappointed it is sooooooo hard to merge such a tiny bugfix. 
The real code change is very small, it keep growing with pytype/pylint/unitest fixes, and finally end up with close.!
